setuptools/distutils_patch.py

```diff
@@ -0,0 +1,31 @@
+"""
+Ensure that the local copy of distutils is preferred over stdlib.
+
+See https://github.com/pypa/setuptools/issues/417#issuecomment-392298401
+for more motivation.
+"""
+
+import sys
+import importlib
+import contextlib
+from os.path import dirname
+
+
+@contextlib.contextmanager
+def patch_sys_path():
+    orig = sys.path[:]
+    sys.path[:] = [dirname(dirname(__file__))]
+    try:
+        yield
+    finally:
+        sys.path[:] = orig
+
+
+def ensure_local_distutils():
+    if 'distutils' in sys.path:
+        raise RuntimeError("Distutils must not be imported before setuptools")
+    with patch_sys_path():
+        importlib.import_module('distutils')
```

Comment:
Apparently this works, since you've tested it, but I'd have expected some import failures once you take the standard library off `sys.path`. Probably by now, everything has been imported already.

An alternative that should not require patching sys.path at all could be:
```python
spec = importlib.util.spec_from_file_location("distutils", PATH_TO_DISTUTILS_INIT_PY)
mod = importlib.util.module_from_spec(spec)
mod.__loader__.exec_module(mod)
```

Reply:
> I'd have expected some import failures once you take the standard library off `sys.path`.

Actually, I would expect this code to work as well when distutils is removed from stdlib, as long as the `distutils` python package is on `sys.path`, which it is in this proposal (it's exposed as a top-level package).

> Probably by now, everything has been imported already.

Can you phrase that in another way. I'm not sure what you're trying to say.



> An alternative that should not require patching sys.path at all could be

Hmm. Yes, perhaps that's better.

Reply:
I tried implementing the alternative, but I don't think `spec_from_file_location` is viable, as that assumes file-system-based packages, and setuptools could be in a zip archive or loaded through another loader.